crates/router/src/core/mandate.rs

```diff
@@ -243,7 +226,80 @@ where
         _ => Some(router_data.request.get_payment_method_data()),
     }
 }
+pub async fn update_mandate_procedure<F, FData>(
+    state: &AppState,
+    resp: types::RouterData<F, FData, types::PaymentsResponseData>,
+    mandate: Mandate,
+    merchant_id: &str,
+    pm_id: Option<String>,
+) -> errors::RouterResult<types::RouterData<F, FData, types::PaymentsResponseData>>
+where
+    FData: MandateBehaviour,
+{
+    let mandate_details = match resp.response.clone() {
+        Ok(types::PaymentsResponseData::TransactionResponse {
+            mandate_reference, ..
+        }) => mandate_reference,
+        _ => None,
+    };
+
+    let new_update_record = payments::UpdateHistory {
+        connector_mandate_id: mandate.connector_mandate_id,
+        payment_method_id: mandate.payment_method_id,
+        original_payment_id: mandate.original_payment_id,
+    };
 
+    let mandate_ref = mandate
+        .connector_mandate_ids
+        .map(|ids| {
+            Some(ids)
+                .parse_value::<payments::ConnectorMandateReferenceId>("Connector Reference Id")
+                .change_context(errors::ApiErrorResponse::MandateDeserializationFailed)
+        })
+        .transpose()?;
+
+    let updated_mandate_ref = match mandate_ref {
+        Some(mut mr) => {
+            match mr.update_history {
+                Some(ref mut uh) => uh.push(new_update_record),
+                None => {
+                    let v = vec![new_update_record];
+                    mr.update_history = Some(v)
+                }
+            };
+            mr.connector_mandate_id = mandate_details
+                .clone()
+                .and_then(|mr| mr.connector_mandate_id);
```

Comment:
Avoid cloning potentially large pieces of data.
```suggestion
            mr.connector_mandate_id = mandate_details
                .as_ref()
                .and_then(|mr| mr.connector_mandate_id.clone());
```